blog: pmemkv's bindings description

lukaszstolarczuk opened the pull request:
preview: https://lukaszstolarczuk.github.io/2020/03/04/pmemkv-bindings.html

before merge:
- [x] pmemkv-python released
- [x] python's and nodejs' docs are available on https://pmem.io

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/pmem.github.io/126)
<!-- Reviewable:end -->

karczex commented:
___
*[_posts/2020-02-15-pmemkv-bindings.md, line 80 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/126#-M02dQ5M3qaz6yf-fJTs:-M0SKNDr-G7j810V3AZW:bozphpd) ([raw file](https://github.com/pmem/pmem.github.io/blob/5516e0b5653c75e2a372f693f8ad8465c331196c/_posts/2020-02-15-pmemkv-bindings.md#L80)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It's not returning numeric value for OK status. Most functions doesn't have return statement at all.  


<!-- Sent from Reviewable.io -->

szyrom commented:
Imho this should be merged after pmemkv-python release AND pmemkv-python blogpost (@karczex ) AND after the merge of nodejs bindings support for direct access to pmem

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+All of our currently released bindings work properly with libpmemkv 1.0. Unfortunately their API
+may not be fully functionally equal to the native C/C++ API. These **bindings, which are
+in version 1.0** deliver API in accordance with **pmemkv 1.0**. In general, pmemkv's binding
+released in version X.Y is (always) functionally greater or equal to pmemkv in the same X.Y version.
```
Drop "in general".

I don't like the "functionally greater or equal". "Is compatible with pmemkv x.y"?

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+may not be fully functionally equal to the native C/C++ API. These **bindings, which are
+in version 1.0** deliver API in accordance with **pmemkv 1.0**. In general, pmemkv's binding
+released in version X.Y is (always) functionally greater or equal to pmemkv in the same X.Y version.
+Additionally, if a binding's version is prior to 1.0 (any 0.x version), it should be considered
```
Bindings with major version 0 should be considered experimental.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+Additionally, if a binding's version is prior to 1.0 (any 0.x version), it should be considered
+an experimental implementation.
+
+Taking into consideration that libpmemkv is developing in the fast pace, there are always
```
"The pmemkv project is currently undergoing heavy development, and it's likely that the core library feature set will outpace features of the various bindings."

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+an experimental implementation.
+
+Taking into consideration that libpmemkv is developing in the fast pace, there are always
+some gaps in the current versions of bindings' implementations. Even more, each binding
```
Moreover, individual bindings are developed independently. We recommend checking with documentation for specific bindings to see if a feature is present.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+is developed in its own pace, so the best way to check compatibility with pmemkv
+is to read changelog of the latest release of a specific binding.
+
+#### Handling pmemkv statuses
```
This paragraph adds nothing to the post. In effect, what you are saying is that "there are some error states, they are handled differently for every language, see docs to check how".

I'd change this to something like "Idiomatic language support" in which you explain that we tried to make bindings feel natural to the language (idiomatic), and because of this we had to slightly divert from the original library in some areas, such as return values, error handling and so on.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+#### Not always direct access to data
+
+**Except in Python** (and soon [in Node.js][nodejs_pr]), our bindings do not (yet)
```
Invert this.

The bindings listed below, unless otherwise noted, do not support ...

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+#### Examples, tests and CI
+
+Each repository is equipped with basic, needed features like:
```
Each repository includes examples, tests and CI.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+#### Packaging specific for language
+
+Each language is different, so we try our best to use packaging and installation systems
```
remove "try our best", just say "we use"

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+#### Contributions are welcome
+
+As all of pmem's organization repositories, the ones of pmemkv-bindings are also
```
Well, it's github, so unless we used a non-free license, it's a given that projects in pmem organization are open source.
I'd rephrase this to say that "pmemkv is BSD, so we welcome contributions."

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+### pmemkv-nodejs
+
+Node.js binding is the first one to be released in stable version (1.0). It means
+it guarantees backward compatibility and it provides API functionally equal to
```
This guarantees backward compatibility and stable API, compatible with libpmemkv 1.0.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+It also delivers API description in form of JSDoc documentation (accessible e.g. as html in
+[here](nodejs_pmem_io)) and NodeJS-native support for configuring engines by JSON object.
+It handles returned statuses as exceptions and does not return any value when no errors occur.
```
It uses exception for error handling.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+It also delivers API description in form of JSDoc documentation (accessible e.g. as html in
+[here](nodejs_pmem_io)) and NodeJS-native support for configuring engines by JSON object.
+It handles returned statuses as exceptions and does not return any value when no errors occur.
+Installation of this project is supported by simply typing `npm install` within the source directory.
```
... uses npm for distribution, and can be installed using with `npm install pmemkv-python`.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+Installation of this project is supported by simply typing `npm install` within the source directory.
+It then installs the dependencies listed in package.json in the local node_modules folder.
+
+For more information, like required dependencies, usage and full installation guide,
```
such as

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+### pmemkv-python
+
+This is the second binding to be released in version 1.0. It has all functionalities
```
Rephrase this to match whats written for nodejs.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+### pmemkv-java
+
+Latest released version is 0.9 and it's working just fine with libpmemkv 1.0, but
```
and is compatible with libpmemkv 1.0, but comes with no guarantees of stable API.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+Latest released version is 0.9 and it's working just fine with libpmemkv 1.0, but
+its API may still change. In this binding's code, the JNI framework is used,
+with implementation of pmemkv's C API delivered in separate repository.
```
This binding uses JNI to interoperate with C, and the pmemkv C JNI implementation is delivered in ...

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+Latest released version is 0.9 and it's working just fine with libpmemkv 1.0, but
+its API may still change. In this binding's code, the JNI framework is used,
+with implementation of pmemkv's C API delivered in separate repository.
+It handles returned statuses as exceptions and does not return any value when no errors occur.
```
same as above.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+### pmemkv-ruby
+
+It's currently released as 0.9 version, properly working with pmemkv 1.0,
```
rephrase this to use the same language as above.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+If you feel like we're missing some language, that may be useful to have in our portfolio,
+or you just want to let us know we don't have information about some (possibly yours)
+pmemkv's binding, or your feel like we're missing some functionalities, don't hesitate to
```
`please file an issue to let us know. We are always eager to help our community grow.` or something like that.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+
+### Looking forward
+
+In future more bindings may come to life. There are currently two (of which we are aware)
```
What are you trying to say?
That we are thinking about creating new bindings or that someone else can create bindings? I'm guessing both, but this isn't very clear to me.

pbalcer commented on _posts/2020-02-15-pmemkv-bindings.md:
```diff
@@ -0,0 +1,155 @@
+### Looking forward
+
+In future more bindings may come to life. There are currently two (of which we are aware)
+still in development and outside of our organization, you can read about them here:
```
What's "outside of our organization"?. Say either "outside of core pmemkv team" or "developed by our community". This is an open source project.

"There are ... community developed bindings".

lukaszstolarczuk commented:
We've merged the node.js PR with direct data access, so I've updated the content already. I've also changed the directory of the post (along with the post date update), so the review is aligned to the old file, sorry for the inconveniences.
